File: suc/kotlin/025.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "026MindMapen"
private const val UI_TYPE = "Productivity"
private const val STYLE_THEME = "Dark Neon"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFF00FFFF)
        val secondary = Color(0xFF00FF88)
        val tertiary = Color(0xFFBB86FC)
        val background = Color(0xFF0D0D0D)
        val surface = Color(0xFF1C1C1C)
        val surfaceVariant = Color(0xFF2E2E2E)
        val outline = Color(0xFF3F3F3F)
        val success = Color(0xFF00FFAA)
        val warning = Color(0xFFFFC107)
        val error = Color(0xFFFF1744)
        val onPrimary = Color(0xFF000000)
        val onSecondary = Color(0xFF000000)
        val onTertiary = Color(0xFF000000)
        val onBackground = Color(0xFFFFFFFF)
        val onSurface = Color(0xFFFFFFFF)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Spacing {
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.12f)
        val level2 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.18f)
    }
}

private val AppColorScheme = darkColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    tertiary = AppTokens.Colors.tertiary,
    onTertiary = AppTokens.Colors.onTertiary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        content = content
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RootScreen() {
    val nodes = remember { mutableStateListOf(Offset(200f, 400f), Offset(500f, 700f), Offset(350f, 250f)) }
    var currentColor by remember { mutableStateOf(AppTokens.Colors.primary) }

    Scaffold(
        contentWindowInsets = WindowInsets(0),
        floatingActionButton = {
            FloatingActionButton(
                onClick = { nodes.add(Offset((100..600).random().toFloat(), (200..800).random().toFloat())) },
                containerColor = AppTokens.Colors.secondary,
                contentColor = AppTokens.Colors.onSecondary,
                shape = CircleShape
            ) { Text("+", style = MaterialTheme.typography.titleMedium) }
        },
        bottomBar = {
            Surface(
                color = AppTokens.Colors.surfaceVariant,
                tonalElevation = AppTokens.ElevationMapping.level1.elevation
            ) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(AppTokens.Spacing.md),
                    horizontalArrangement = Arrangement.SpaceEvenly,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Button(
                        onClick = { currentColor = AppTokens.Colors.primary },
                        colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary, contentColor = AppTokens.Colors.onPrimary)
                    ) { Text("Cyan") }
                    Button(
                        onClick = { currentColor = AppTokens.Colors.secondary },
                        colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.secondary, contentColor = AppTokens.Colors.onSecondary)
                    ) { Text("Green") }
                    Button(
                        onClick = { currentColor = AppTokens.Colors.tertiary },
                        colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.tertiary, contentColor = AppTokens.Colors.onTertiary)
                    ) { Text("Purple") }
                }
            }
        },
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text("Neon Mind Map", style = MaterialTheme.typography.displayLarge, color = AppTokens.Colors.primary) },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors(containerColor = AppTokens.Colors.background)
            )
        },
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(AppTokens.Colors.background)
                .pointerInput(Unit) {
                    detectDragGestures { change, _ ->
                        nodes.add(change.position)
                    }
                }
        ) {
            Canvas(modifier = Modifier.fillMaxSize()) {
                for (i in 1 until nodes.size) {
                    drawLine(
                        color = currentColor,
                        start = nodes[i - 1],
                        end = nodes[i],
                        strokeWidth = 6f
                    )
                }
                nodes.forEach {
                    drawCircle(color = currentColor, radius = 20f, center = it)
                }
            }
        }
    }


}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFF0D0D0D)
@Composable
fun PreviewScreen() {
    AppTheme {
        RootScreen()
    }
}

File: suc/swift/025.swift

import SwiftUI

// MARK: - Hex Color Helper
// 扩展 Color 以支持十六进制颜色初始化，方便与 Kotlin 代码对应
extension Color {
    init(hex: UInt) {
        self.init(
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0
        )
    }
}

// MARK: - AppTokens
// 对应 Kotlin 的 AppTokens 对象，定义了颜色、字体、间距和阴影规范
struct AppTokens {
    struct Colors {
        static let primary = Color(hex: 0xFF00FFFF)
        static let secondary = Color(hex: 0xFF00FF88)
        static let tertiary = Color(hex: 0xFFBB86FC)
        static let background = Color(hex: 0xFF0D0D0D)
        static let surface = Color(hex: 0xFF1C1C1C)
        static let surfaceVariant = Color(hex: 0xFF2E2E2E)
        static let outline = Color(hex: 0xFF3F3F3F)
        static let success = Color(hex: 0xFF00FFAA)
        static let warning = Color(hex: 0xFFFFC107)
        static let error = Color(hex: 0xFFFF1744)
        static let onPrimary = Color(hex: 0xFF000000)
        static let onSecondary = Color(hex: 0xFF000000)
        static let onTertiary = Color(hex: 0xFF000000)
        static let onBackground = Color(hex: 0xFFFFFFFF)
        static let onSurface = Color(hex: 0xFFFFFFFF)
    }

    struct TypographyTokens {
        // Kotlin 的 FontWeight.Normal 对应 Swift 的 .regular
        static let display = Font.system(size: 28, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular)
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Spacing {
        // Dp 单位直接映射为 CGFloat
        static let sm: CGFloat = 8
        static let md: CGFloat = 12
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
    }

    struct ShadowSpec {
        let elevation: CGFloat
        let radius: CGFloat
        let dy: CGFloat
        let opacity: Double // Kotlin 的 Float 对应 Swift 的 Double
    }

    struct ElevationMapping {
        static let level1 = ShadowSpec(elevation: 2, radius: 4, dy: 2, opacity: 0.12)
        static let level2 = ShadowSpec(elevation: 6, radius: 8, dy: 4, opacity: 0.18)
    }
}

// MARK: - RootScreen
// 对应 Kotlin 的 RootScreen Composable 函数
struct RootScreen: View {
    // 对应 Kotlin 的 mutableStateListOf 和 mutableStateOf
    @State private var nodes: [CGPoint] = [CGPoint(x: 200, y: 400), CGPoint(x: 500, y: 700), CGPoint(x: 350, y: 250)]
    @State private var currentColor: Color = AppTokens.Colors.primary

    var body: some View {
        ZStack {
            // 背景色，忽略安全区域以实现全屏
            AppTokens.Colors.background
                .ignoresSafeArea(.all)

            // Canvas 用于绘制节点和连线
            Canvas { context, size in
                // 绘制连线
                for i in 1..<nodes.count {
                    let start = nodes[i - 1]
                    let end = nodes[i]
                    var path = Path()
                    path.move(to: start)
                    path.addLine(to: end)
                    context.stroke(path, with: .color(currentColor), lineWidth: 6) // strokeWidth = 6f
                }

                // 绘制节点
                for node in nodes {
                    var path = Path()
                    // radius = 20f, 所以直径为 40f
                    path.addEllipse(in: CGRect(x: node.x - 20, y: node.y - 20, width: 40, height: 40))
                    context.fill(path, with: .color(currentColor))
                }
            }
            // 对应 Kotlin 的 pointerInput + detectDragGestures
            .gesture(
                DragGesture(minimumDistance: 0) // minimumDistance: 0 确保每次位置变化都触发
                    .onChanged { value in
                        nodes.append(value.location)
                    }
            )
            .ignoresSafeArea(.all) // Canvas 也忽略安全区域，确保绘制内容覆盖整个屏幕

            // 顶部栏 (Top Bar)
            VStack {
                Text("Neon Mind Map")
                    .font(AppTokens.TypographyTokens.display)
                    .foregroundColor(AppTokens.Colors.primary)
                    // 垂直内边距，使其高度与 Compose 的 TopAppBar 视觉上接近
                    .padding(.vertical, AppTokens.Spacing.lg)
                    .frame(maxWidth: .infinity) // 填充宽度
                    .background(AppTokens.Colors.background)
                Spacer() // 将顶部栏推到顶部
            }
            .ignoresSafeArea(.container, edges: .top) // 顶部栏延伸到屏幕顶部边缘

            // 底部栏 (Bottom Bar)
            VStack {
                Spacer() // 将底部栏推到底部
                HStack(spacing: AppTokens.Spacing.md) { // 按钮之间的间距
                    // Cyan 按钮
                    Button(action: { currentColor = AppTokens.Colors.primary }) {
                        Text("Cyan")
                            .font(AppTokens.TypographyTokens.body)
                            .foregroundColor(AppTokens.Colors.onPrimary)
                            .padding(.horizontal, AppTokens.Spacing.lg) // 按钮内容水平内边距
                            .padding(.vertical, AppTokens.Spacing.sm) // 按钮内容垂直内边距
                    }
                    .background(AppTokens.Colors.primary)
                    .cornerRadius(15) // 根据内容高度 (14pt + 8pt*2 = 30pt)，圆角设为 15pt 形成胶囊形
                    .buttonStyle(PlainButtonStyle()) // 移除 SwiftUI 默认按钮样式

                    // Green 按钮
                    Button(action: { currentColor = AppTokens.Colors.secondary }) {
                        Text("Green")
                            .font(AppTokens.TypographyTokens.body)
                            .foregroundColor(AppTokens.Colors.onSecondary)
                            .padding(.horizontal, AppTokens.Spacing.lg)
                            .padding(.vertical, AppTokens.Spacing.sm)
                    }
                    .background(AppTokens.Colors.secondary)
                    .cornerRadius(15)
                    .buttonStyle(PlainButtonStyle())

                    // Purple 按钮
                    Button(action: { currentColor = AppTokens.Colors.tertiary }) {
                        Text("Purple")
                            .font(AppTokens.TypographyTokens.body)
                            .foregroundColor(AppTokens.Colors.onTertiary)
                            .padding(.horizontal, AppTokens.Spacing.lg)
                            .padding(.vertical, AppTokens.Spacing.sm)
                    }
                    .background(AppTokens.Colors.tertiary)
                    .cornerRadius(15)
                    .buttonStyle(PlainButtonStyle())
                }
                .padding(AppTokens.Spacing.md) // 整个按钮行的内边距
                .frame(maxWidth: .infinity) // 填充宽度
                .background(AppTokens.Colors.surfaceVariant)
                // 对应 Kotlin 的 tonalElevation 和 ShadowSpec
                .shadow(color: .black.opacity(AppTokens.ElevationMapping.level1.opacity),
                        radius: AppTokens.ElevationMapping.level1.radius,
                        x: 0,
                        y: AppTokens.ElevationMapping.level1.dy)
            }
            .ignoresSafeArea(.container, edges: .bottom) // 底部栏延伸到屏幕底部边缘

            // 浮动操作按钮 (Floating Action Button)
            VStack {
                Spacer() // 将 FAB 推到底部
                HStack {
                    Spacer() // 将 FAB 推到右侧
                    Button(action: {
                        // 对应 Kotlin 的随机位置生成
                        let randomX = CGFloat.random(in: 100...600)
                        let randomY = CGFloat.random(in: 200...800)
                        nodes.append(CGPoint(x: randomX, y: randomY))
                    }) {
                        Text("+")
                            .font(AppTokens.TypographyTokens.title)
                            .foregroundColor(AppTokens.Colors.onSecondary)
                            .frame(width: 56, height: 56) // 标准 FAB 尺寸，假设 56dp
                            .background(AppTokens.Colors.secondary)
                            .clipShape(Circle()) // 圆形
                            // 对应 Kotlin 的 ShadowSpec
                            .shadow(color: .black.opacity(AppTokens.ElevationMapping.level2.opacity),
                                    radius: AppTokens.ElevationMapping.level2.radius,
                                    x: 0,
                                    y: AppTokens.ElevationMapping.level2.dy)
                    }
                    .padding(AppTokens.Spacing.lg) // FAB 距离屏幕边缘的内边距
                }
            }
        }
        .statusBarHidden(true) // 隐藏状态栏，对应 Kotlin 的 WindowInsetsControllerCompat.hide(WindowInsetsCompat.Type.systemBars())
    }
}

// MARK: - App Entry Point
// 对应 Kotlin 的 MainActivity 和 setContent
@main
struct MindMapApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
        }
    }
}